### PR TITLE
Extend MySQL and PostgreSQL connectors

### PR DESCRIPTION
**Pull request by bonifacyj**
**- What I did**

1. Added table size, in MB, to the Tag Template 
2. Implemented scraping of the table_size for MySQL and PostgreSQL
3. Implemented scraping of create_time and update_time for MySQL tables
4. Introduced user config (see "ingest_cnfg.yaml"), where a user can toggle options for updating metadata and scraping additional metadata
5. Implemented metadata update with ANALYZE command for PostgreSQL andMySQL
6. Implemented timing of the ANALYZE command for user to know how long it takes
7. Implemented (optional) scraping of num_rows in MySQL and PostgreSQL tables

**- How I did it**

According to design doc

**- How to verify it**

1. Create two databases: MySQL and PostgreSQL
2. Populate them with random data via metadata generator ([Postgres](https://github.com/mesmacosta/google-datacatalog-connectors-profiler/tree/master/postgresql/metadata_generator), [MySQL](https://github.com/mesmacosta/google-datacatalog-connectors-profiler/tree/master/mysql/metadata_generator)) 
2.1 (Optional) To get number of rows counted, add some data to the newly created tables, as the metadata generation script leaves the tables empty.
3. Set up a project on GCP.
4. Run connectors ([Postgres](https://github.com/GoogleCloudPlatform/datacatalog-connectors-rdbms/tree/master/google-datacatalog-postgresql-connector), [MySQL](https://github.com/GoogleCloudPlatform/datacatalog-connectors-rdbms/tree/master/google-datacatalog-mysql-connector)) 
4.1 You can try running them with different user configurations. Configurations are stored in ingest_cfg.yaml file. 
5. In DataCatalog, Tag Template "XY Table - Metadata", you should see a new field - "table_size_MB". In Table Tags you should see value for this new field. If there are any data in your tables, you should see the num_rows field of Table Tags populated as well. Besides, there should be timestamps in 'Created' and 'Last Modified' fields of entry details for MySQL tables.

**- Description for the changelog**
Extend PostgreSQL and MySQL connectors

### CODE REVIEW

File: google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py [hunk 1]
```diff
@@ -77,6 +87,93 @@ def _get_metadata_from_rdbms_connection(self, connection_args, query):
...
+        con = None
+        try:
+            con = self._create_rdbms_connection(connection_args)
+            cur = con.cursor()
+            start_update = time.time()
+            for query in update_queries:
+                self._execute_update_query(cur, query)
+            end_update = time.time()
+            logging.info(
+                'Metadata update took {} seconds to run.'
```
Comment (socha23) [inline google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py:126]: 'recalculating metadata' or 'metadata analysis' perhaps? The point of the message is for the user to understand what's going on.

File: google-datacatalog-mysql-connector/ingest_cnfg.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# information whether column nullable
+# column data type,
+# max length of char values in a column
+# and numeric precision of values in a column.
+#
+# Using this configuration file, you can turn on or off the scraping of other pieces of data.
+
+# Runs ANALYZE command for each user table, to update metadata tables and make database statistics more accurate.
+# Note, that it might take long time to ANALYZE big tables
+update-metadata: true
```
[the remark this answers is not in the document]
Comment (bonifacyj) [reply google-datacatalog-mysql-connector/ingest_cnfg.yaml:20]: Regarding case: noted;
regarding "run_analyze_table" - I'd like to have something more universal, because ANALYZE is not in SQL standard. So if in some DB the command not called ANALYZE, we still can use this same option in config file, without confusing the user. What about "refresh_metadata_info" or "refresh_metadata_tables"?

File: google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/scrape/query_assembler.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import os
+from google.datacatalog_connectors.rdbms.scrape import query_assembler
+
+
+class QueryAssembler(query_assembler.QueryAssembler):
+
+    def _get_update_statement(self, table_name):
```
[the remark this answers is not in the document]
Comment (bonifacyj) [reply google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/scrape/query_assembler.py]: > also will this work for all legit table names without any quoting / escaping?

I think so. Works for all the test data I've tried it with. Do you have some specific example in mind with which we can try maybe?

File: google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/config/num_rows_query.sql [hunk 4]
```diff
@@ -0,0 +1,9 @@
+SELECT  t.table_schema as database_name,
+        t.table_name, t.table_rows
+
+      FROM information_schema.tables t
+
+      WHERE t.table_schema NOT IN
+          ('mysql', 'information_schema',
```
Comment (mesmacosta) [inline google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/config/num_rows_query.sql:7]: Do you think it makes sense to extract those internal system tables names to a constant or another config file?

the other query has the same internal system tables.

Comment (bonifacyj) [reply google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/config/num_rows_query.sql:7]: I find it's a good idea :) but I don't know how to do it in SQL. Does it have sort of tuple/list data structure where I can store multiple strings?

Comment (mesmacosta) [reply google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/config/num_rows_query.sql:7]: You can use the SQL file as a template SQL, and create the resolved SQL in the code.

I remember that you were working on a Query Assembler, the Query Assembler could resolve the SQL file.

if we are going to need multiple queries, and they all will have those same internal system tables, probably worth doing that.

If you are going to send other Pull Requests, I'd suggest doing this when you need another query, for the other fields you will be working on... So this Pull Request is not too big.

Comment (bonifacyj) [reply google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/config/num_rows_query.sql:7]: for now, this is all I've got in terms of PRs :) I'll do some more work on MySQL/PostrgreSQL connectors at the end of my internship, if there is time left. Let's then postpone it until that moment in time.

File: google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py [hunk 5]
```diff
@@ -86,3 +186,18 @@ def _create_rdbms_connection(self, connection_args):
     @classmethod
     def _get_metadata_from_csv(cls, csv_path):
         return pd.read_csv(csv_path)
+
+    def _get_query_assembler(self):
```
Comment (mesmacosta) [inline google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py:190]: `_get_query_assembler` and `_execute_refresh_query` are required if you are using the query user config.

Since the rdbms is a commons library, I think this should be in the rdbms README.md for guidance on implementing new RDBMS connectors. WDYT?

File: google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py [hunk 6]
```diff
@@ -77,6 +87,96 @@ def _get_metadata_from_rdbms_connection(self, connection_args, query):
...
+                if len(rows) == 0:
+                    warnings.warn(
+                        "Query {} delivered no rows. Skipping it.".format(
+                            query))
+                else:
+                    new_dataframe = self._create_dataframe(rows)
+                    new_dataframe.columns = [
+                        item[0].lower() for item in cur.description
+                    ]
+                    base_dataframe = self._get_merged_dataframe(
```
Comment (mesmacosta) [inline google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py:160]: rename this variable?

This contains a wrong meaning should be merged_dataframe, no?

File: google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py [hunk 7]
```diff
@@ -77,6 +87,96 @@ def _get_metadata_from_rdbms_connection(self, connection_args, query):
...
+
+    def _get_optional_metadata_from_rdbms_connection(self, connection_args,
+                                                     optional_queries,
+                                                     base_dataframe,
+                                                     metadata_definition):
+        con = None
+        try:
+            con = self._create_rdbms_connection(connection_args)
+            cur = con.cursor()
+            for query in optional_queries:
```
Comment (mesmacosta) [inline google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py:148]: If you have a list of queries, and one of them throws an error... this seems to be hard to troubleshoot.

Can you log some information about  the query that will be executed at the beginning of the for iteration?

**Review by mesmacosta — CHANGES_REQUESTED** [review]

File: google-datacatalog-mysql-connector/README.md [hunk 8]
```diff
@@ -127,9 +127,14 @@ export MYSQL2DC_RAW_METADATA_CSV=mysql_raw_csv (If supplied ignores the MYSQL se
 
 ```
 
-## 3. Run entry point
+## 3. Adapt user configurations
 
-### 3.1. Run Python entry point
+Along with default metadata, the connector can ingest optional metadata as well, such as number of
+rows in each table. Please check configuration file ingest_cfg.yaml and adapt according to your needs.
```
[the remark this answers is not in the document]
Comment (bonifacyj) [reply google-datacatalog-mysql-connector/README.md:133]: Maybe it makes sense to just move the file to ../config folder then and adapt the calls in my code? So that it is easier for users to just use it, without having to create their own config. And if they want to  change it, they can just go to source and make changes in the file and save?.. or it doesn't work that way?

So do I understand correctly, that if we install the package via PyPI, the root folder would be postgresql/ ?

File: google-datacatalog-mysql-connector/ingest_cfg.yaml [hunk 9]
```diff
@@ -0,0 +1,23 @@
+---
+# By default, the connector retrieves the following metadata from a MySQL database:
```
[the remark this answers is not in the document]
Comment (bonifacyj) [reply google-datacatalog-mysql-connector/ingest_cfg.yaml:2]: I like this idea, it totally makes sense. I'll do it that way then.

File: google-datacatalog-mysql-connector/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (mesmacosta) [reply google-datacatalog-mysql-connector/README.md:133]: When you install via PyPI, it's going to be installed in their python library folder, like:
`env/lib/python3.7/site-packages/google/datacatalog_connectors/rdbms`.

So this directory is not user friendly, they would need to change the config there.

I like the way the file is being loaded using: `os.getcwd()`. So this way it's going to load the file where the connector is being executed. So we just need to instruct users to have the config file at same execution dir. WDYT?

File: google-datacatalog-postgresql-connector/src/google/datacatalog_connectors/postgresql/config/metadata_query.sql [hunk 10]
```diff
@@ -36,10 +36,13 @@ SELECT t.table_schema as schema_name,
        END as column_type,
        c.character_maximum_length as column_char_length,
        c.numeric_precision as column_numeric_precision,
-       e.enum_values as column_enum_values
+       e.enum_values as column_enum_values,
+       CAST (pg_total_relation_size(pc.oid) AS FLOAT) / 1024 / 1024 as table_size
```
[the remark this answers is not in the document]
Comment (bonifacyj) [reply google-datacatalog-postgresql-connector/src/google/datacatalog_connectors/postgresql/config/metadata_query.sql:40]: doesn't look to be the case. I reviewed the documentation again, and there is no mentioning of pg_total_relation_size() returning NULL or something that is not integer. I've also tested with empty tables and empty views, and in this case it returns 0.

File: google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/config/metadata_query.sql [hunk 11]
```diff
@@ -16,6 +16,8 @@
 
 SELECT  t.table_schema as database_name,
         t.table_name, t.table_type,
+        t.create_time, t.update_time,
+        round((t.data_length+t.index_length) / (1024 * 1024), 3) as table_size,
```
[the remark this answers is not in the document]
Comment (bonifacyj) [reply google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/config/metadata_query.sql:20]: Regarding the query - I checked again, and t.data_length and t.index_length could be NULL if a table is in fact a view. But then query doesn't fail, it just returns NULL as a table_size. So if I remember correctly, if that happens, then the tag field "Table size" will not be populated with anything.

Regarding variable name, you are right, it totally makes sense to rename it.

File: google-datacatalog-rdbms-connector/tests/google/datacatalog_connectors/rdbms/scrape/query_assembler_test.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+import os
+import unittest
+
+from .. import test_utils
+from google.datacatalog_connectors.commons_test import utils
+from google.datacatalog_connectors.rdbms.scrape import config
+import mock
+
+
+class QueryAssemblerTestCase(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (bonifacyj) [reply google-datacatalog-rdbms-connector/tests/google/datacatalog_connectors/rdbms/scrape/query_assembler_test.py:26]: Yes, totally! I see what you mean

File: google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py (hunk 5 above)
Comment (bonifacyj) [reply google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py:196]: Sure, I'll add this information

Comment (bonifacyj) [reply google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/scrape/metadata_scraper.py:196]: I looked at it now, and I'm thinking it actually makes sense to add some context about configurable queries in general into the README.

File: google-datacatalog-mysql-connector/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (bonifacyj) [reply google-datacatalog-mysql-connector/README.md]: Alright, please have a look at the README if it's clear now.

[the remark this answers is not in the document]
Comment (mesmacosta) [reply google-datacatalog-mysql-connector/README.md]: LGTM

**Review by mesmacosta — APPROVED** [review]

**Comment by mesmacosta** [comment]
LGTM, just one final thing @bonifacyj, were you able to run the 1000 tables scenario, with the `ANALYSE` statement.

Just to see if it makes sense to update the metrics file: https://github.com/GoogleCloudPlatform/datacatalog-connectors-rdbms/tree/master/google-datacatalog-mysql-connector/docs/README.md.

**Comment by bonifacyj** [comment]
> LGTM, just one final thing @bonifacyj, were you able to run the 1000 tables scenario, with the `ANALYSE` statement?
> 
> Just to see if it makes sense to update the metrics file: https://github.com/GoogleCloudPlatform/datacatalog-connectors-rdbms/tree/master/google-datacatalog-mysql-connector/docs/README.md.

No, let me do that and tell you the results. It will definitely take longer than usual, I'll report to which extend

**Comment by bonifacyj** [comment]
> > LGTM, just one final thing @bonifacyj, were you able to run the 1000 tables scenario, with the `ANALYSE` statement?
> > Just to see if it makes sense to update the metrics file: https://github.com/GoogleCloudPlatform/datacatalog-connectors-rdbms/tree/master/google-datacatalog-mysql-connector/docs/README.md.
> 
> No, let me do that and tell you the results. It will definitely take longer than usual, I'll report to which extend

It's strange, but I got 27,9 mins. Strange because in the metrics without ANALYZE it says it also takes approximately 27 mins.

**Comment by mesmacosta** [comment]
> > > LGTM, just one final thing @bonifacyj, were you able to run the 1000 tables scenario, with the `ANALYSE` statement?
> > > Just to see if it makes sense to update the metrics file: https://github.com/GoogleCloudPlatform/datacatalog-connectors-rdbms/tree/master/google-datacatalog-mysql-connector/docs/README.md.
> > 
> > 
> > No, let me do that and tell you the results. It will definitely take longer than usual, I'll report to which extend
> 
> It's strange, but I got 27,9 mins. Strange because in the metrics without ANALYZE it says it also takes approximately 27 mins.

The execution time will depend on the environment, network and execution. The network might be the most decisive factor, since the connector does multiple API calls, and now SQL statements when the flag is enabled.

27,9 mins takes close to the old execution time, so that's good enough. If you want to run it with the refresh_statement disabled to compare, but no need to.

So just waiting for the FIX to we mentioned internally to merge this.

**Review by mesmacosta — APPROVED** [review]
